File: output/xCodeEval_8000/C++/C/29d33ec3efff45ee9f12b18e7461d439.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_SIZE 1000000

typedef char i64;
typedef long double ld;
typedef char* vs;
typedef char* vc;
typedef int bool;
typedef struct {
    i64 first;
    i64 second;
} pii;
typedef struct {
    i64 first;
    pii second;
} pip;
typedef struct {
    pii first;
    i64 second;
} ppi;

const long long Mod = 1000000007LL, INF = 1e9, LINF = 1e18;
const long double Pi = 3.141592653589793116, EPS = 1e-9, Gold = ((1+sqrt(5))/2);
long long keymod[] = {1000000007LL, 1000000009LL, 1000000021LL, 1000000033LL};
long long keyCount = sizeof(keymod) / sizeof(long long);

int getbit(i64 s, int i) { return (s >> i) & 1; }
i64 onbit(i64 s, int i) { return s | (i64(1) << i); }
i64 offbit(i64 s, int i) { return s & (~(i64(1) << i)); }
int cntbit(i64 s) { return __builtin_popcountll(s); }

typedef struct {
    int argc;
    char** argv;
} ControlIOArgs;

void ControlIO(ControlIOArgs args);
void TimerStart();
void TimerStop();
void Exit();
char* cppstr_infile = "FILE.IN";
char* cppstr_outfile = "FILE.OUT";

typedef struct {
    pii first;
    pii second;
} ppp;
char s1[MAX_SIZE], s2[MAX_SIZE], virus[MAX_SIZE];
i64 n1, n2, nv;
int NextCommonPrefix[MAX_SIZE][26];

void ProceedVirus() {
    for (i64 i=0; i<nv; i++) {
        for (i64 j=0; j<26; j++) {
            char pref2[MAX_SIZE];
            strncpy(pref2, virus, i);
            pref2[i] = (char)(65 + j);
            pref2[i+1] = '\0';
            for (i64 x=i+1; x>=0; x--) {
                if (strncmp(pref2+(i+1-x), virus, x) == 0) {
                    NextCommonPrefix[i][j] = x;
                    break;
                }
            }
        }
    }
}

void Input() {
    scanf("%s %s %s", s1, s2, virus);
    nv = strlen(virus);
    n1 = strlen(s1);
    n2 = strlen(s2);
}

void Solve() {
    int List1[26][MAX_SIZE], List2[26][MAX_SIZE];
    for (i64 i=0; i<n1; i++) {
        List1[s1[i]-'A'][i] = i;
    }
    for (i64 i=0; i<n2; i++) {
        List2[s2[i]-'A'][i] = i;
    }
    ProceedVirus();
    ppp Default = { {-1, -1}, {-1, -1} };
    bool dp[MAX_SIZE][MAX_SIZE][MAX_SIZE][MAX_SIZE];
    ppp Last[MAX_SIZE][MAX_SIZE][MAX_SIZE][MAX_SIZE];
    for (i64 i=0; i<n1; i++) {
        for (i64 j=0; j<n2; j++) {
            if (s1[i] == s2[j]) {
                dp[i][j][1][NextCommonPrefix[0][s1[i]-'A']] = 1;
            } else {
                dp[i][j][0][0] = 1;
            }
        }
    }
    for (i64 x=1; x<MIN(n1, n2); x++) {
        for (i64 i=0; i<n1-1; i++) {
            for (i64 j=0; j<n2-1; j++) {
                for (i64 y=0; y<nv; y++) {
                    if (!dp[i][j][x][y]) continue;
                    for (i64 z=0; z<26; z++) {
                        if (List1[z][i] == -1) continue;
                        if (List2[z][j] == -1) continue;
                        i64 i1 = List1[z][i];
                        i64 j1 = List2[z][j];
                        dp[i1][j1][x+1][NextCommonPrefix[y][z]] = 1;
                        Last[i1][j1][x+1][NextCommonPrefix[y][z]] = (ppp) { {i, j}, {x, y} };
                    }
                }
            }
        }
    }
    for (i64 x=MIN(n1, n2); x>0; x--) {
        for (i64 y=0; y<nv; y++) {
            for (i64 i=0; i<n1; i++) {
                for (i64 j=0; j<n2; j++) {
                    if (!dp[i][j][x][y]) continue;
                    i64 I = i, J = j, X = x, Y = y;
                    char s[MAX_SIZE];
                    i64 idx = 0;
                    while (I != -1) {
                        s[idx++] = s1[I];
                        ppp Token = Last[I][J][X][Y];
                        I = Token.first.first;
                        J = Token.first.second;
                        X = Token.second.first;
                        Y = Token.second.second;
                    }
                    s[idx] = '\0';
                    printf("%s\n", s);
                    return;
                }
            }
        }
    }
    printf("0\n");
}

int main(int argc, char* argv[]) {
    ControlIO((ControlIOArgs) {argc, argv});
    
    Input();
    TimerStart();
    Solve();
    TimerStop();
    
    return 0;
}

void ControlIO(ControlIOArgs args) {
    char* infile = (char*)malloc(strlen(cppstr_infile)+1);
    char* outfile = (char*)malloc(strlen(cppstr_outfile)+1);
    strcpy(infile, cppstr_infile);
    strcpy(outfile, cppstr_outfile);
    #ifdef Akikaze
    time_t t = time(0);
    struct tm* now = localtime(&t);
    printf("Source code by #Team4T-Akikaze.\n");
    printf("Current time: ");
    printf("%d-%s%d-%s%d | %s%d:%s%d:%s%d\n\n", (now->tm_year + 1900),
        ((now->tm_mon < 9) ? "0" : ""), (now->tm_mon + 1),
        ((now->tm_mday < 10) ? "0" : ""), now->tm_mday,
        ((now->tm_hour < 10) ? "0" : ""), now->tm_hour,
        ((now->tm_min < 10) ? "0" : ""), now->tm_min,
        ((now->tm_sec < 10) ? "0" : ""), now->tm_sec);
    
    #ifdef OImode
    printf("OI-Mode: ON\n");
    printf("Input file: %s\n", infile);
    printf("Output file: %s\n\n", outfile);
    #else
    printf("OI-Mode: OFF\n");
    printf("Input file: NULL\n");
    printf("Output file: NULL\n\n");
    #endif
    
    #ifdef MultiTest
    printf("MultiTest-Mode: ON\n");
    #else
    printf("MultiTest-Mode: OFF\n");
    #endif
    
    #ifdef Interactive
    printf("Interactive-Mode: ON\n\n");
    #else
    printf("Interactive-Mode: OFF\n\n");
    #endif
    
    if (args.argc > 1) freopen(args.argv[1], "r", stdin);
    if (args.argc > 2) freopen(args.argv[2], "w", stdout);
    #elif OImode
    freopen(infile, "r", stdin);
    freopen(outfile, "w", stdout);
    #endif
}

void TimerStart() {
    #ifdef Akikaze
    TimeStart = clock();
    #endif
}

void TimerStop() {
    #ifdef Akikaze
    TimeEnd = clock();
    double ElapsedTime = (double)(TimeEnd - TimeStart) / CLOCKS_PER_SEC;
    printf("\n\nTime elapsed: %lf seconds.\n", ElapsedTime);
    #endif
}

void Exit() {
    TimerStop();
    exit(0);
}

File: output/xCodeEval_8000/C++/Java/29d33ec3efff45ee9f12b18e7461d439.java
import java.util.*;
import java.io.*;

class Main {
    static final int Mod = 1000000007, INF = 1_000_000_000, LINF = 1_000_000_000_000_000L;
    static final double Pi = 3.141592653589793116, EPS = 1e-9, Gold = ((1 + Math.sqrt(5)) / 2);
    static final long[] keymod = {1000000007L, 1000000009L, 1000000021L, 1000000033L};
    static final int keyCount = keymod.length;

    static int getbit(long s, int i) {
        return (int) ((s >> i) & 1);
    }

    static long onbit(long s, int i) {
        return s | (1L << i);
    }

    static long offbit(long s, int i) {
        return s & (~(1L << i));
    }

    static int cntbit(long s) {
        return Long.bitCount(s);
    }

    static long TimeStart, TimeEnd;
    static final int OImode = 227420978, MultiTest = 227420978, Interactive = 227420978;

    static void ControlIO(String[] args) {
        String cppstr_infile = "FILE.IN";
        String cppstr_outfile = "FILE.OUT";

        String infile = cppstr_infile;
        String outfile = cppstr_outfile;

        #ifdef Akikaze
        System.out.println("Source code by #Team4T-Akikaze.");
        System.out.println("Current time: " + new SimpleDateFormat("yyyy-MM-dd | HH:mm:ss").format(new Date()) + "\n");

        #ifdef OImode
        System.out.println("OI-Mode: ON");
        System.out.println("Input file: " + infile);
        System.out.println("Output file: " + outfile);
        System.out.println();
        #else
        System.out.println("OI-Mode: OFF");
        System.out.println("Input file: NULL");
        System.out.println("Output file: NULL");
        System.out.println();
        #endif

        #ifdef MultiTest
        System.out.println("MultiTest-Mode: ON");
        #else
        System.out.println("MultiTest-Mode: OFF");
        #endif

        #ifdef Interactive
        System.out.println("Interactive-Mode: ON\n");
        #else
        System.out.println("Interactive-Mode: OFF\n");
        #endif

        if (args.length > 0) {
            try {
                System.setIn(new FileInputStream(args[0]));
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        #elif OImode
        try {
            System.setIn(new FileInputStream(infile));
            System.setOut(new PrintStream(new FileOutputStream(outfile)));
        } catch (Exception e) {
            e.printStackTrace();
        }
        #endif
    }

    static void TimerStart() {
        #ifdef Akikaze
        TimeStart = System.nanoTime();
        #endif
    }

    static void TimerStop() {
        #ifdef Akikaze
        TimeEnd = System.nanoTime();
        double ElapsedTime = (TimeEnd - TimeStart) / 1e9;
        System.out.println("\n\nTime elapsed: " + ElapsedTime + " seconds.");
        #endif
    }

    static void Exit() {
        TimerStop();
        System.exit(0);
    }

    static void proceedVirus(String virus, int nv, int[][] nextCommonPrefix) {
        for (int i = 0; i < nv; i++) {
            for (int j = 0; j < 26; j++) {
                StringBuilder pref2 = new StringBuilder(virus.substring(0, i));
                pref2.append((char) (65 + j));
                for (int x = i + 1; x >= 0; x--) {
                    if (pref2.substring(i + 1 - x, i + 1).equals(virus.substring(0, x))) {
                        nextCommonPrefix[i][j] = x;
                        break;
                    }
                }
            }
        }
    }

    static void solve() {
        Scanner sc = new Scanner(System.in);
        String s1 = sc.next();
        String s2 = sc.next();
        String virus = sc.next();
        int n1 = s1.length();
        int n2 = s2.length();
        int nv = virus.length();
        int[][] nextCommonPrefix = new int[nv][26];

        proceedVirus(virus, nv, nextCommonPrefix);

        List<Integer>[] list1 = new List[26];
        List<Integer>[] list2 = new List[26];
        for (int i = 0; i < 26; i++) {
            list1[i] = new ArrayList<>();
            list2[i] = new ArrayList<>();
        }

        for (int i = 0; i < n1; i++) {
            list1[s1.charAt(i) - 'A'].add(i);
        }

        for (int i = 0; i < n2; i++) {
            list2[s2.charAt(i) - 'A'].add(i);
        }

        boolean[][][][] dp = new boolean[n1][n2][Math.min(n1, n2) + 1][nv + 1];
        int[][][][][] last = new int[n1][n2][Math.min(n1, n2) + 1][nv + 1][4];
        for (int i = 0; i < n1; i++) {
            for (int j = 0; j < n2; j++) {
                if (s1.charAt(i) == s2.charAt(j)) {
                    dp[i][j][1][nextCommonPrefix[0][s1.charAt(i) - 'A']] = true;
                } else {
                    dp[i][j][0][0] = true;
                }
            }
        }

        for (int x = 1; x < Math.min(n1, n2); x++) {
            for (int i = 0; i < n1 - 1; i++) {
                for (int j = 0; j < n2 - 1; j++) {
                    for (int y = 0; y < nv; y++) {
                        if (!dp[i][j][x][y]) {
                            continue;
                        }
                        for (int z = 0; z < 26; z++) {
                            if (upperBound(list1[z], i) == list1[z].size()) {
                                continue;
                            }
                            if (upperBound(list2[z], j) == list2[z].size()) {
                                continue;
                            }
                            int i1 = list1[z].get(upperBound(list1[z], i));
                            int j1 = list2[z].get(upperBound(list2[z], j));
                            dp[i1][j1][x + 1][nextCommonPrefix[y][z]] = true;
                            last[i1][j1][x + 1][nextCommonPrefix[y][z]][0] = i;
                            last[i1][j1][x + 1][nextCommonPrefix[y][z]][1] = j;
                            last[i1][j1][x + 1][nextCommonPrefix[y][z]][2] = x;
                            last[i1][j1][x + 1][nextCommonPrefix[y][z]][3] = y;
                        }
                    }
                }
            }
        }

        for (int x = Math.min(n1, n2); x > 0; x--) {
            for (int y = 0; y < nv; y++) {
                for (int i = 0; i < n1; i++) {
                    for (int j = 0; j < n2; j++) {
                        if (!dp[i][j][x][y]) {
                            continue;
                        }
                        int I = i, J = j, X = x, Y = y;
                        StringBuilder s = new StringBuilder();
                        while (I != -1) {
                            s.insert(0, s1.charAt(I));
                            int[] token = last[I][J][X][Y];
                            I = token[0];
                            J = token[1];
                            X = token[2];
                            Y = token[3];
                        }
                        System.out.println(s);
                        return;
                    }
                }
            }
        }

        System.out.println("0");
    }

    public static void main(String[] args) {
        ControlIO(args);

        #ifndef Interactive
        Scanner sc = new Scanner(System.in);
        solve();
        #else
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt();
        while (T-- > 0) {
            solve();
        }
        #endif
    }
}
